design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1067 @@
+
+_Phase 2 should be completed by release of Crossplane v0.14_
+
+Related Issues:
+- [Package dependency resolution
+  design](https://github.com/crossplane/crossplane/issues/434)
+- [Check required fields and better error messages for Package
+  Install](https://github.com/crossplane/crossplane/issues/896)
+
+The main purpose of Phase 2 is to establish the `PackageLock` type and implement
+the building blocks for dependency management. However, there is no intention to
+resolve dependencies for packages at this stage. Instead, users will be provided
+more granular information about how they can manually resolve dependencies to
+make the installation of a package successful. For example, in Phase 1, when a
+user creates a `Package` a `PackageRevision` will always be created and it will
+try to install everything it brings. If it is unsuccessful, the `Package` will
+reflect that in its `status`, but will be no more granular than specifying which
+`PackageRevision` is failing and perhaps what operation it is failing on.
+
+In Phase 2, the `Package` controller will instead check the current set of
+packages in the cluster and refuse to create the `PackageRevision` if there is a
+known violation. To do so, the `Package` controller must be able to read and
+write the state of the packages it manages in a "concurrency-safe" manner. Let's
+look at an example.
+
+**Package A (PA) and Package B (PB) Both Want to Install CRD1 and CRD2**
+
+In the model established in Phase 1, this scenario could result in one of three
+outcomes:
+
+1. PA successfully installs and controls CRD1 and CRD2. It starts its packaged
+   controller and runs without issue.
+2. PB successfully installs and controls CRD1 and CRD2. It starts its packaged
+   controller and runs without issue.
+3. PB successfully installs and controls CRD1, but before it can install CRD2,
+   PA installs and controls it. Neither one of the package installations are
+   successful and none of the packaged controllers are started.
+
+While each of the controllers in this scenario are acting in a deterministic
+manner, the observed behavior is not consistent. We cannot alter this situation
+such that only PA or only PB will always win out and install successfully, but
+we can ensure that one or the other will install successfully (barring other
+violations).
+
+### PackageLock API
+
+The `PackageLock` type is a cluster-scoped singleton resource type that keeps a
+record of ownership and dependencies of Crossplane packages.
+
+```go
+// A PackageLock keeps track of Crossplane packages.
+type PackageLock struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   PackageLockSpec   `json:"spec,omitempty"`
+  Status PackageLockStatus `json:"status,omitempty"`
+}
+
+type PackageLockSpec struct {
+  Packages                  []PackageDependencies `json:"packages,omitempty"`
+  Compositions              map[string]string     `json:"compositions,omitempty"`
```

Comment:
@negz it appears your link suggests this does not comply. However, I also read through [this linked issue](https://github.com/kubernetes/kubernetes/issues/2004) and I think it's possible that this situation would be acceptable because theses fields may not be viewed as "subobjects". In practice, the reason for using a map here is that the common operation we want to perform is checking whether a resource already is installed and owned by another package, which is much more efficient with maps. What are your thoughts?

Reply:
I'm happy with them being maps as long as we can justify that in relation to API conventions. My link was intended to make sure you were aware of the convention rather than to suggest this was in violation. I think my "what do these contain" question stands though. Does the design explain what goes in these somewhere? e.g. Is it a map of GVK to package name or something?

Reply:
Yep! For compositions it would be composition name to package name and for CRDs it would be GVK to package name.